spec/index.bs

```diff
@@ -7364,14 +7380,40 @@ rendering operations have been flushed to the drawing buffer.
         that created this swap chain next.
 
         <div algorithm="GPUSwapChain.getCurrentTexture">
-            **Called on:** {{GPUSwapChain}} this.
+            **Called on:** {{GPUSwapChain}} |this|.
 
             **Returns:** {{GPUTexture}}
 
-            Issue: Describe {{GPUSwapChain/getCurrentTexture()}} algorithm steps.
+            1. Set |this|.{{GPUSwapChain/[[dirty]]}} to `true`.
+            1. Return |this|.{{GPUSwapChain/[[currentTexture]]}}.
         </div>
 </dl>
 
+<div algorithm>
+    During the "update the rendering [of the] `Document`" step of the "[=Update the rendering=]"
+    HTML processing model, the associated {{HTMLCanvasElement}} |canvas| for each {{GPUSwapChain}}
+    |swapChain| should <dfn abstract-op>present the swap chain content to the canvas</dfn> by running
+    the following steps:
+
+        1. If |swapChain|.{{GPUSwapChain/[[dirty]]}} is `false`, stop.
```

Comment:
I'm a bit confused by this. We have this RAF callback, and we are supposed to present a frame at the end of it.
At the same time, the `[[currentTexture]]` is an internal property of the swap chain, which changes on each new RAF. So what constitutes an invalidation of the frame here? It seems like `getCurrentTexture` is not an action that should affect the frame, in which case, if the user doesn't call it, the frame should still be auto-cleared as all resources are. But then, `dirty` seems unnecessary?

Reply:
I've addressed everything else, but I'd like to discuss this a bit more.

First off, you're right that `[[dirty]]` is unnecessary. I can set `[[currentTexture]]` to `null` until `getCurrentTexture()` is called again and that has the same effect. I've made that change.

As far as what constitutes invalidation of the frame, I disagree that the frame should be auto-cleared. We've seen from multiple WebGL applications in the past that the ability to render a single frame and then leave it displayed until, say, the user interacts with the canvas, is valuable both for page performance and battery life. Examples of pages that utilize this feature off the top of my head include Google Maps and Sketchfab.

Because of WebGL's implicit default backbuffer the canvas invalidation logic in that API was unfortunately complicated:

> WebGL presents its drawing buffer to the HTML page compositor immediately before a compositing operation, but only if at least one of the following has occurred since the previous compositing operation:
> 
> Context creation
> Canvas resize
> clear, drawArrays, or drawElements has been called while the drawing buffer is the currently bound framebuffer

Given that we have a more explicit swap chain model here it seems like a good opportunity to simplify. I would argue that there's not much point to calling `getCurrentTexture()` unless you're planning on rendering, so that seems like a perfect, low-impact point at which to flag that the canvas has been invalidated. Otherwise we need to start tracking whether or not a texture produced by a swap chain is bound as a color attachment to any render passes which were submitted to a queue, an also whether or not the texture was used as a copy destination in a `copyBufferToTexture` or `copyTextureToTexture` call, and I'm not sure if compute shaders can alter textures but... well you get the idea. :)

Reply:
+1, I completely agree that getCurrentTexture in a new frame should be what clears the canvas.

Reply:
> I would argue that there's not much point to calling `getCurrentTexture()` unless you're planning on rendering, so that seems like a perfect, low-impact point at which to flag that the canvas has been invalidated.

> +1, I completely agree that getCurrentTexture in a new frame should be what clears the canvas.

The name "getCurrentTexture()" implies to any reader a const/immutable and idempotent function. If this starts to have active logic tied to it, the name would be good to change? Maybe `acquireNextTexture()` or something similar? Having a "getter" that has active behavior is brittle.

The reasoning that there is no much point to calling `getCurrentTexture()` unless you're planning on rendering does make sense, but as projects grow more complex, as does logic, and given the innocent name, people can think "well there's no harm calling that getter function when I'm not rendering, since it oughta be fast anyway", and do not realize they need to optimize away calls to it to keep the previous frame visible.

Reply:
Talked about this in the editors meeting, we're going to propose to the group replacing `getCurrentTexture()` with an `acquireNextTexture()` which can only be called once per frame (throws error if called again in the same frame).

Thinking about it a little more now, we need to think about whether it could end up being fragile. E.g. rendering is done off mouse events instead of rAF, and you get more than one mouse event in a frame (it can happen!)

Reply:
Some old minutes links
[2019-02-04](https://docs.google.com/document/d/1HXmaPNE7zzp-mzjbZRHLz_sXjpwS78IYovqNLGaS_wY/edit#heading=h.dpg974h3hno4) talking about getNextTexture+present -> getCurrentTexture
[2019-05-16 F2F Mountain View](https://docs.google.com/document/d/1HR9vPk48-lhzHAsyF57QzjKMmi0gQHfSXMmyqZuFWkc/edit#heading=h.3sus85csnbov) talking about imagebitmap swap chain
I thought we discussed in San Diego (misspoke as Seattle) but doesn't appear so.

Reply:
Moved discussion to https://github.com/gpuweb/gpuweb/issues/1424